dcp_prototype/backend/ledger/code/common/ledger_orm.py

```diff
@@ -24,161 +24,232 @@ def session(self, **kwargs):
+    updated_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
 
 
 class SequencingProtocol(Base):
     __tablename__ = "sequencing_protocol"
-    id = Column(String(), primary_key=True)
-    paired_end_sequencing = Column(String(), nullable=True)
-    instrument_manufacturer_model = Column(String(), nullable=True)
-    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
-    updated_at = Column(
-        DateTime, default=datetime.utcnow, nullable=False, onupdate=datetime.utcnow
+
+    id = Column(String, primary_key=True)
+    paired_end_sequencing = Column(String)
+    instrument_manufacturer_model = Column(String)
+    created_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+    updated_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+
+
+class AnalysisFileQuantificationProtocolExpressionFileProcessJ(Base):
+    __tablename__ = "analysis_file_quantification_protocol_expression_file_process_j"
+
+    id = Column(String, primary_key=True)
+    analysis_file_id = Column(ForeignKey("analysis_file.id"), nullable=False)
+    quantification_protocol_id = Column(
+        ForeignKey("quantification_protocol.id"), nullable=False
     )
+    expression_file_id = Column(ForeignKey("expression_file.id"), nullable=False)
+    created_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+    updated_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+
+    analysis_file = relationship("AnalysisFile")
+    expression_file = relationship("ExpressionFile")
+    quantification_protocol = relationship("QuantificationProtocol")
+
+
+class BamQcMetric(Base):
+    __tablename__ = "bam_qc_metric"
+
+    id = Column(String, primary_key=True)
+    analysis_file_id = Column(ForeignKey("analysis_file.id"), nullable=False)
+    percent_mapped_to_genome = Column(String)
+    percent_reads_mapped_to_intergenic = Column(String)
+    percent_reads_mapped_uniquely = Column(String)
+    percent_reads_mapped_multiple = Column(String)
+    created_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+    updated_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+
+    analysis_file = relationship("AnalysisFile")
+
+
+class Library(Base):
+    __tablename__ = "library"
+
+    id = Column(String, primary_key=True)
+    project_id = Column(ForeignKey("project.id"), nullable=False)
+    created_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+    updated_at = Column(DateTime(True), nullable=False, server_default=text("now()"))
+
+    project = relationship("Project")
 
-    libraries = relationship(
-        "Library",
-        order_by=Library.id,
-        back_populates="sequencing_protocol",
-        cascade="all, delete, delete-orphan",
+
+class ProjectContributorJoin(Base):
+    __tablename__ = "project_contributor_join"
+
+    id = Column(String, primary_key=True)
```

Comment:
General question about the id on the join tables. Based on a few examples I saw in sqlalchemy modeling many-to-many relationships, it seems like what the thing to do is have the things that are being joined, jointly be the primary key. For example, for this table, it would be `contributor_id` and `project_id` together. Then no separate `id` is specified. 

This begs the question about which join tables in the metadata schema actually have biological value and which are simply there to connect the entities because that should influence whether the table gets a unique id or not and influenced whether I create a unique HCA-accession for it or not.

Reply:
@mckinsel ^^

Reply:
There is no "right" way to handle ids in join tables. They can be an auto incrementing integer. They can be a join of the two ids. They can be a separate accession id. I don't have an opinion here and can change it to whatever makes most sense from the data wrangling point of view